LLama/Native/NativeLogConfig.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+
+namespace LLama.Native;
+
+/// <summary>
+/// Configure llama.cpp logging
+/// </summary>
+public class NativeLogConfig
+{
+    /// <summary>
+    /// Callback from llama.cpp with log messages
+    /// </summary>
+    /// <param name="level"></param>
+    /// <param name="message"></param>
+    public delegate void LLamaLogCallback(LLamaLogLevel level, string message);
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+    [DllImport(NativeApi.libraryName, CallingConvention = CallingConvention.Cdecl, EntryPoint = "llama_log_set")]
+    private static extern void native_llama_log_set(LLamaLogCallback? logCallback);
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logCallback"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(LLamaLogCallback? logCallback)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        if (NativeLibraryConfig.LibraryHasLoaded)
+        {
+            // The library is loaded, just pass the callback directly to llama.cpp
+            native_llama_log_set(logCallback);
+        }
+        else
+        {
+            // We can't set the log method yet since that would cause the llama.dll to load.
+            // Instead configure it to be set when the native library loading is done
+            NativeLibraryConfig.Instance.WithLogCallback(logCallback);
+        }
+    }
+
+    /// <summary>
+    /// Register a callback to receive llama log messages
+    /// </summary>
+    /// <param name="logger"></param>
+#pragma warning disable IDE1006 // Naming Styles (name imitates llama.cpp)
+    public static void llama_log_set(ILogger? logger)
+#pragma warning restore IDE1006 // Naming Styles
+    {
+        // Clear the logger
+        if (logger == null)
+        {
+            llama_log_set((LLamaLogCallback?)null);
+            return;
+        }
+
+        // Bind a function that converts into the correct log level
+        llama_log_set((level, message) =>
+        {
+            logger.Log(level.ToLogLevel(), "(llama.cpp): {message}", message);
+        });
+    }
```

Comment:
Looks like this is slightly inconsistent with NativeLibraryConfig.WithLogCallback(ILogger) when the method called before the initialization: it saves a callback while WithLogCallback saves a logger and NativeApi.Log has different logic for this two cases.

Reply:
You mean because the `ILogger` is used at the top of the `Log` method, but the callback in only used after the filtering? For now I can fix that by moving the callback up to the top as well.

I think this probably needs a better fix though. Perhaps just removing the default console logging altogether, so you _must_ supply some kind of logger if you want log messages.

Reply:
> You mean because the ILogger is used at the top of the Log method, but the callback in only used after the filtering?

Yes, not only the filtering, but also the `(llama.cpp): ` prefix here that will be absent in direct `ILogger` call.

I guess this consistency problem could be solved by "duplicating" llama_log_set(LLamaLogCallback? logCallback) logic, but for ILogger.

Or maybe we should not support `ILogger` here and in `WithLogCallback` at all for now. This make sense for low level messages from llama.cpp, because more sophisticated ways of handling them should be applied. For example, llama.cpp calls a log callback for each `.` symbol in `...........` line that is printed during the loading, and a hundred `.` log messages is not ideal for ILogger.

Reply:
I've changed this to redirect `WithLogCallback(ILogger? logger)` to `NativeLogConfig.llama_log_set` which will wrap it in a delegate and call back into `WithLogCallback(NativeLogConfig.LLamaLogCallback? callback)`. That way the formatting is consistent and there's no duplicated logic.

I also removed all of the default printing to console, that allows all the associated things (enable logging, log level) to be removed.

> For example, llama.cpp calls a log callback for each `.` symbol in `...........` line....

I have noticed that, any thoughts on how it could be better handled?

Reply:
> I also removed all of the default printing to console, that allows all the associated things (enable logging, log level) to be removed.

That's fine by me, but not sure this is friendly to users who use console. Maybe it is better to embrace llama.cpp strategy: write to the console always unless there is a callback provided? Then LLamaSharp behavior will be fully consistent with llama.cpp.

Reply:
I've added in a system to buffer together messages until a newline is encountered, as you suggested. At the moment I've only done this for the `ILogger` method, not the direct callback. I don't want to add too much "magic" over the base function here.